src/mlpack/methods/ann/rbm/spike_slab_rbm_policy_impl.hpp

```diff
@@ -0,0 +1,336 @@
+
+#ifndef MLPACK_METHODS_ANN_RBM_SPIKE_SLAB_RBM_POLICY_IMPL_HPP
+#define MLPACK_METHODS_ANN_RBM_SPIKE_SLAB_RBM_POLICY_IMPL_HPP
+
+#include "spike_slab_rbm_policy.hpp"
+
+namespace mlpack {
+namespace ann {
+template<typename DataType>
+SpikeSlabRBMPolicy<DataType>::SpikeSlabRBMPolicy(
+    const size_t visibleSize,
+    const size_t hiddenSize,
+    const size_t poolSize,
+    const ElemType slabPenalty,
+    ElemType radius):
+    visibleSize(visibleSize),
+    hiddenSize(hiddenSize),
+    poolSize(poolSize),
+    slabPenalty(slabPenalty),
+    radius(2 * radius)
+{
+  parameter.set_size(visibleSize * hiddenSize * poolSize +
+      visibleSize + hiddenSize);
+
+  visibleMean.set_size(visibleSize, 1);
+  spikeMean.set_size(hiddenSize, 1);
+  spikeSamples.set_size(hiddenSize, 1);
+  slabMean.set_size(poolSize, hiddenSize);
+};
+
+// Reset function
+template<typename DataType>
+void SpikeSlabRBMPolicy<DataType>::Reset()
+{
+  // Weight shape D * K * N
+  weight = arma::Cube<ElemType>(parameter.memptr(),
```

Comment:
Do you think it might be possible to use ``arma::mat`` instead of ``arma::cube`` for the weights, that way we could vectorize some expressions (avoid a for loop), since you can't e.g. multiply a cube with a matrix and store the result as matrix. It looks like there are a lot of places that would come in handy.

Reply:
Well, I just saw the code again. Most of the function's actually use `weight.slice(i)`
so I think we would not be able to vectorize most of the expressions. We could use submat something like this `weight.submat(i * visibleSize * poolSize, 0, (i + 1) * visibleSize * poolSize , poolSize)` that would not get rid of the loop but i don't if armadillo matrix is preferred over cubes for performance considerations.

Reply:
I'll take a second look, and let you know if I find an expression that could benefit from ``arma::mat`` over ``arma::cube``, in the meantime, can you test different numbers of iterations?